File: solution/23MergekSortedLists.cpp
//
// Created by Alaric on 2019-09-26.
//
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode* dummy = new ListNode(0);
        ListNode* tail = dummy;

        auto comp = [](ListNode* a , ListNode* b) {return a->val > b->val;};
        priority_queue<ListNode* , vector<ListNode*>, decltype(comp)> q(comp);

        for(ListNode* list:lists )
            if (list) q.push(list);

        while(!q.empty()){
            tail->next = q.top();
            q.pop();
            tail = tail->next;
            if (tail->next) q.push(tail->next);
        }
        return dummy->next;
    }
};

File: solution/23MergekSortedLists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        import heapq
        dummy = ListNode(0)
        tail = dummy
        queue = []
        for i in range(len(lists)):
            if lists[i] != None:
                heapq.heappush(queue , (lists[i].val , i))
                lists[i] = lists[i].next
        while queue:
            val,index = heapq.heappop(queue)
            tail.next = ListNode(val)
            tail = tail.next
            if lists[index]:
                heapq.heappush(queue , (lists[index].val , index))
                lists[index] = lists[index].next
        return dummy.next
